service-rpc/client/esdm_rpc_get_random_bytes_full_timeout.c

```diff
@@ -0,0 +1,136 @@
+ * DAMAGE.
+ */
+
+#include <errno.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/types.h>
+
+#include "esdm_rpc_client_helper.h"
+#include "esdm_rpc_client_internal.h"
+#include "esdm_rpc_service.h"
+#include "math_helper.h"
+#include "esdm_logger.h"
+#include "ptr_err.h"
+#include "ret_checkers.h"
+#include "visibility.h"
+
+struct esdm_get_random_bytes_full_timeout_buf {
+	ssize_t ret;
+	uint8_t *buf;
+	size_t buflen;
+};
+
+static void
+esdm_rpcc_get_random_bytes_full_timeout_cb(
+	const GetRandomBytesFullTimeoutResponse *response,
+	void *closure_data)
+{
+	struct esdm_get_random_bytes_full_timeout_buf *buffer =
+		(struct esdm_get_random_bytes_full_timeout_buf *)closure_data;
+
+	esdm_rpcc_error_check(response, buffer);
+
+	if (response->ret < 0) {
+		buffer->ret = response->ret;
+		return;
+	}
+
+	buffer->ret = (ssize_t)min_size(response->randval.len, buffer->buflen);
+	memcpy(buffer->buf, response->randval.data, (size_t)buffer->ret);
+
+	/* Zeroization of response is handled in esdm_rpc_client_read_handler */
+}
+
+static bool esdm_time_after(time_t curr, time_t base)
+{
+	if (curr == (time_t)-1)
+		return false;
+	if (base == (time_t)-1)
+		return true;
+	return (curr > base) ? true : false;
+}
+
+DSO_PUBLIC
+ssize_t esdm_rpcc_get_random_bytes_full_timeout_int(uint8_t *buf, size_t buflen,
+						    struct timespec *ts,
+						    void *int_data)
+{
+	GetRandomBytesFullTimeoutRequest msg =
+		GET_RANDOM_BYTES_FULL_TIMEOUT_REQUEST__INIT;
+	esdm_rpc_client_connection_t *rpc_conn = NULL;
+	struct esdm_get_random_bytes_full_timeout_buf buffer;
+	size_t maxbuflen = buflen, orig_buflen = buflen;
+	time_t base;
+	ssize_t ret = 0;
+
+	CKNULL(ts, -EINVAL);
+
+	CKINT(esdm_rpcc_get_unpriv_service(&rpc_conn, int_data));
+
+	base = time(NULL);
+	base += ts->tv_sec;
+
+	while (buflen) {
+		buffer.ret = -ETIMEDOUT;
+		buffer.buf = buf;
+		buffer.buflen = buflen;
+
+		msg.len = min_size(maxbuflen, buflen);
+		msg.tv_sec = (uint64_t)ts->tv_sec;
+		msg.tv_nsec = (uint32_t)ts->tv_nsec;
+
+		unpriv_access__rpc_get_random_bytes_full_timeout(
+			&rpc_conn->service, &msg,
+			esdm_rpcc_get_random_bytes_full_timeout_cb, &buffer);
+
+		if (buffer.ret < -255) {
+			maxbuflen = (size_t)(-buffer.ret);
+			continue;
+		} else if (buffer.ret == -EAGAIN) {
+			time_t curr = time(NULL);
+
+			/* Terminate the endless loop if the timeout hits */
+			if (esdm_time_after(curr, base))
+				break;
+
+			nanosleep(&esdm_client_poll_ts, NULL);
+			continue;
+		} else if (buffer.ret < 0) {
+			ret = buffer.ret;
+			goto out;
+		}
+
+		esdm_test_shm_status_add_rpc_client_written((size_t)buffer.ret);
+		buflen -= (size_t)buffer.ret;
+		buf += buffer.ret;
+	}
+
+out:
+	esdm_rpcc_put_unpriv_service(rpc_conn);
+	return (ret < 0) ? ret : (ssize_t)orig_buflen;
+}
+
+DSO_PUBLIC
+ssize_t esdm_rpcc_get_random_bytes_full_timeout(uint8_t *buf, size_t buflen,
+						struct timespec *ts)
+{
+	return esdm_rpcc_get_random_bytes_full_timeout_int(buf, buflen, NULL,
```

Comment:
The parameters for timespec and int_data are in the wrong order, causing -EINVAL on every call through esdm_rpcc_get_random_bytes_full_timeout

Reply:
After fixing this issues first tests looked good, the timeout handling was working correctly and I also did some tests with concurrent calls to that functions.